internal/providers/vault/provider.go

```diff
@@ -39,53 +42,71 @@ func (p *Provider) GetName() string {
-// to know which field to return from the map. By default, it returns the 'value'.
-// An alternative field can be obtained by appending '#fieldName' to the id argument.
-func (p *Provider) GetValue(id string) (value []byte, err error) {
-	id, fieldName := parseVaultID(id)
-
-	var secret *vault.Secret
-	if secret, err = p.Client.Logical().Read(id); err != nil {
-		return
-	}
-	// secret can be nil if it's not found
-	if secret == nil {
-		err = fmt.Errorf("HashiCorp Vault provider could not find a secret called '%s'", id)
-		return
+// parseSecret returns value navigated by given fields on secret object.
+// Note that a secret returned from Vault is effectively a JSON object.
+func parseSecret(secret *vault.Secret, path string, fields []string) ([]byte, error) {
+	// navigate to secret value
+	var obj, ok = secret.Data, false
+	for _, field := range fields[:len(fields)-1] {
+		obj, ok = obj[field].(map[string]interface{})
+		if !ok {
+			err := fmt.Errorf("HashiCorp Vault provider expects the secret '%s' to contain field '%s' along '%s'", path, field, strings.Join(fields, "."))
+			return nil, err
+		}
 	}
 
-	var ok bool
-	var valueObj interface{}
-	valueObj, ok = secret.Data[fieldName]
-	if !ok {
-		err = fmt.Errorf("HashiCorp Vault provider expects the secret '%s' to contain field '%s'", id, fieldName)
-		return
-	}
+	var field = fields[len(fields)-1]
+	var value = obj[field]
 
-	switch v := valueObj.(type) {
+	// secret value must be either string or bytes
+	switch v := value.(type) {
 	case string:
-		value = []byte(v)
+		return []byte(v), nil
 	case []byte:
-		value = v
+		return v, nil
 	default:
-		err = fmt.Errorf("HashiCorp Vault provider expects the secret to be a string or byte[], got %T", v)
+		err := fmt.Errorf("HashiCorp Vault provider expects the secret to be a string or byte[], got %T", v)
+		return nil, err
+	}
+}
+
+// GetValue obtains a value by id. The id should contain the path in Vault to the secret. It may be appended with a
+// hash following the object property path to the secret value; defaults to DefaultField.
+// For example:
+//   - `kv/database/password` returns the value of field `value` in the secret object at given path.
+//   - `kv/database#password` returns the value of field `password` in the secret object at path `kv/database`.
+//   - `secret/data/database#data.value` returns the value of field `value` wrapped in object `data` in secret object
+//  	at path `secret/data/database`.
+// Secrets in Vault are stored as (JSON) objects in the shape of map[string]interface{}. Both path to the secret and
+// fields to the value in the secret must follow Vault API client conventions. Please see documentation of Vault for
+// details.
+func (p *Provider) GetValue(id string) ([]byte, error) {
+	path, fields := parseVaultID(id)
+
+	var err error
+	var secret *vault.Secret
+	if secret, err = p.Client.Logical().Read(path); err != nil {
```

Comment:
You might be able to make this a bit cleaner with use of `:=`
```suggestion
	secret, err := p.Client.Logical().Read(path)
	if err != nil {
```